osquery/tables/system/windows/windows_eventlog.cpp

```diff
@@ -0,0 +1,237 @@
+
+  row["time"] = INTEGER(windows_event.osquery_time);
+  row["datetime"] = SQL_TEXT(windows_event.datetime);
+  row["channel"] = SQL_TEXT(windows_event.source);
+  row["provider_name"] = SQL_TEXT(windows_event.provider_name);
+  row["provider_guid"] = SQL_TEXT(windows_event.provider_guid);
+  row["eventid"] = INTEGER(windows_event.event_id);
+  row["task"] = INTEGER(windows_event.task_id);
+  row["level"] = INTEGER(windows_event.level);
+  row["pid"] = INTEGER(windows_event.pid);
+  row["tid"] = INTEGER(windows_event.tid);
+
+  row["keywords"] = SQL_TEXT(windows_event.keywords);
+  row["data"] = SQL_TEXT(windows_event.data);
+
+  if (context.hasConstraint("time_range", EQUALS)) {
+    auto time_range = context.constraints["time_range"].getAll(EQUALS);
+    row["time_range"] = SQL_TEXT(*time_range.begin());
+  } else {
+    row["time_range"] = SQL_TEXT("");
+  }
+
+  if (context.hasConstraint("timestamp", EQUALS)) {
+    auto timestamp = context.constraints["timestamp"].getAll(EQUALS);
+    row["timestamp"] = SQL_TEXT(*timestamp.begin());
+  }
+
+  if (context.hasConstraint("xpath", EQUALS)) {
+    auto xpaths = context.constraints["xpath"].getAll(EQUALS);
+    row["xpath"] = SQL_TEXT(*xpaths.begin());
+  }
+
+  yield(std::move(row));
+}
+
+void parseQueryResults(QueryContext& context,
+                       EVT_HANDLE queryResults,
+                       RowYield& yield) {
+  std::vector<EVT_HANDLE> events(kNumEventsBlock);
+  unsigned long numEvents = 0;
+
+  // Retrieve the events one block at a time
+  auto ret = EvtNext(
+      queryResults, kNumEventsBlock, events.data(), INFINITE, 0, &numEvents);
+
+  while (ret != FALSE) {
+    for (unsigned long i = 0; i < numEvents; i++) {
+      unsigned long renderedBuffSize = 0;
+      unsigned long renderedBuffUsed = 0;
+      unsigned long propCount = 0;
+      if (!EvtRender(nullptr,
+                     events[i],
+                     EvtRenderEventXml,
+                     renderedBuffSize,
+                     nullptr,
+                     &renderedBuffUsed,
+                     &propCount)) {
+        if (GetLastError() != ERROR_INSUFFICIENT_BUFFER) {
+          LOG(WARNING) << "Failed to allocated memory to render event "
+                       << GetLastError();
+          continue;
+        }
+      }
+
+      std::vector<wchar_t> renderedContent(renderedBuffUsed);
+      renderedBuffSize = renderedBuffUsed;
+      if (!EvtRender(nullptr,
+                     events[i],
+                     EvtRenderEventXml,
+                     renderedBuffSize,
+                     renderedContent.data(),
+                     &renderedBuffUsed,
+                     &propCount)) {
+        LOG(WARNING) << "Failed to render windows event with "
+                     << GetLastError();
+        continue;
+      }
+
+      std::wstringstream xml_event;
+      xml_event << renderedContent.data();
+      parseWelXml(context, xml_event.str(), yield);
+      EvtClose(events[i]);
```

Comment:
It looks like there are a couple of cases above where EvtClose could be missed.

Reply:
Thanks @zwass for the review. I updated handling of `EvtClose`. In case of error such as `RPC_S_INVALID_BOUND` the events will get closed with the query handler.